File: models/lampshade/surface.mjs
const TAU = Math.PI * 2;
export function geometryKey(p) {
  const previewOnly = new Set([
    "density_mode",
    "bulb_length",
    "bulb_offset",
    "bulb_gap",
    "bulb_overlay",
  ]);
  return JSON.stringify(
    Object.fromEntries(
      Object.entries(p).filter(([key]) => !previewOnly.has(key)),
    ),
  );
}
export function profileRadius(p, t) {
  const u = t < 0.5 ? t * 2 : t * 2 - 1;
  const b = u * (1 - p.curve) + ((1 - Math.cos(Math.PI * u)) / 2) * p.curve;
  return (
    ((t < 0.5 ? p.bottom_diameter : p.middle_diameter) * (1 - b) +
      (t < 0.5 ? p.middle_diameter : p.top_diameter) * b) /
    2
  );
}
let cacheKey = "",
  cache;
export function surfacePlan(p) {
  const key = JSON.stringify([
    p.height,
    p.bottom_diameter,
    p.middle_diameter,
    p.top_diameter,
    p.curve,
    p.lobes,
    p.ripple_depth,
    p.twist,
    p.density,
    p.cell_pitch,
    p.cell_aspect,
    p.thickness,
  ]);
  if (key === cacheKey) return cache;
  const nz = 128,
    na = 1024,
    arcs = [],
    lengths = [];
  for (let j = 0; j <= nz; j++) {
    const r = profileRadius(p, j / nz),
      arc = [0];
    for (let i = 0; i < na; i++) {
      const a = (TAU * (i + 0.5)) / na,
        phase = p.lobes * a;
      arc.push(
        arc[i] +
          (Math.hypot(
            r + p.ripple_depth * Math.sin(phase),
            p.ripple_depth * p.lobes * Math.cos(phase),
          ) *
            TAU) /
            na,
      );
    }
    lengths.push(arc[na]);
    arcs.push(arc.map((v) => v / arc[na]));
  }
  function angle(u, z) {
    const t = Math.max(0, Math.min(nz, (z / p.height) * nz)),
      j = Math.min(nz - 1, Math.floor(t)),
      f = t - j;
    u = ((u % 1) + 1) % 1;
    const at = (i) => arcs[j][i] * (1 - f) + arcs[j + 1][i] * f;
    let lo = 0,
      hi = na;
    while (hi - lo > 1) {
      const m = (hi + lo) >> 1;
      if (at(m) < u) lo = m;
      else hi = m;
    }
    return (TAU * (lo + (u - at(lo)) / (at(hi) - at(lo)))) / na;
  }
  function point(u, z) {
    const a = angle(u, z) + (((p.twist * Math.PI) / 180) * z) / p.height,
      r =
        profileRadius(p, z / p.height) +
        p.ripple_depth *
          Math.sin(
            p.lobes * (a - (((p.twist * Math.PI) / 180) * z) / p.height),
          );
    return [r * Math.cos(a), r * Math.sin(a), z];
  }
  const vertical = [0];
  for (let j = 1; j <= nz; j++) {
    let d = 0;
    for (let i = 0; i < 64; i++) {
      const a = point(i / 64, (p.height * (j - 1)) / nz),
        b = point(i / 64, (p.height * j) / nz);
      d += Math.hypot(...a.map((v, k) => v - b[k]));
    }
    vertical.push(vertical[j - 1] + d / 64);
  }
  const circumference =
    (lengths.reduce((a, b) => a + b, 0) - (lengths[0] + lengths[nz]) / 2) / nz;
  const height = vertical[nz],
    pitch = Math.max(
      p.cell_pitch || 32 - (26 * p.density) / 100,
      2 * p.thickness + 2,
    );
  const cols = Math.max(8, Math.round(circumference / pitch)),
    rows = Math.max(3, Math.round(height / (pitch * p.cell_aspect)));
  function z(v) {
    const target = Math.max(0, Math.min(1, v)) * height;
    let lo = 0,
      hi = nz;
    while (hi - lo > 1) {
      const m = (hi + lo) >> 1;
      if (vertical[m] < target) lo = m;
      else hi = m;
    }
    return (
      (p.height *
        (lo + (target - vertical[lo]) / (vertical[hi] - vertical[lo]))) /
      nz
    );
  }
  cacheKey = key;
  return (cache = {
    circumference,
    maxCircumference: Math.max(...lengths),
    minCircumference: Math.min(...lengths),
    height,
    pitch,
    cols,
    rows,
    width: circumference / cols,
    cellHeight: height / rows,
    angle,
    z,
  });
}
export function bulbClearance(p) {
  const mount = p.height - p.mount_depth - p.plate_thickness;
  const top = mount - p.bulb_offset,
    bottom = top - p.bulb_length;
  const lo = Math.max(0, bottom),
    hi = Math.min(p.height, top);
  if (
    p.fit_test ||
    lo >= hi ||
    Object.values(p).some((v) => !Number.isFinite(v))
  )
    return { diameter: 0, bottom, top, available: false };
  let r = Infinity;
  for (let i = 0; i <= 256; i++) {
    const z = lo + ((hi - lo) * i) / 256;
    r = Math.min(
      r,
      profileRadius(p, z / p.height) -
        (p.lobes ? p.ripple_depth : 0) -
        p.thickness / 2,
    );
  }
  for (const [a, b] of [
    [0, 3],
    [p.height - 3, p.height],
    [mount - 6, mount - 3],
  ])
    if (hi >= a && lo <= b) {
      for (let i = 0; i <= 32; i++) {
        const z =
          Math.max(lo, a) + ((Math.min(hi, b) - Math.max(lo, a)) * i) / 32;
        r = Math.min(
          r,
          profileRadius(p, z / p.height) -
            (p.lobes ? p.ripple_depth : 0) -
            p.thickness -
            1.3,
        );
      }
    }
  if (hi >= mount - 8 && lo <= mount) r = Math.min(r, p.clamp_diameter / 2);
  if (hi >= mount && lo <= mount + p.plate_thickness)
    r = Math.min(r, (p.fixture_diameter + p.hole_clearance) / 2);
  const diameter = Math.max(0, 2 * (r - p.bulb_gap));
  return {
    diameter,
    bottom,
    top,
    available: true,
    protrudes: bottom < 0 || top > p.height,
  };
}

File: models/lampshade/surface.rs
//! Equal-distance coordinates on the shade's centre surface.
use super::*;
const NZ: usize = 128;
const NA: usize = 1024;
pub struct Surface {
    arcs: Vec<Vec<f64>>,
    vertical: Vec<f64>,
    pub circumference: f64,
    pub min_circumference: f64,
    pub max_circumference: f64,
    pub height: f64,
    pub horizontal_step: f64,
    axial_height: f64,
}
impl Surface {
    pub fn new(p: &Params) -> Self {
        let mut arcs = vec![];
        let mut max_curvature = 0f64;
        let mut lengths = vec![];
        for j in 0..=NZ {
            let r = profile(p, j as f64 / NZ as f64);
            let mut arc = vec![0.];
            for i in 0..NA {
                let a = TAU * (i as f64 + 0.5) / NA as f64;
                let phase = p["lobes"] * a;
                let radius = r + p["ripple_depth"] * phase.sin();
                let slope = p["ripple_depth"] * p["lobes"] * phase.cos();
                let second = -p["ripple_depth"] * p["lobes"].powi(2) * phase.sin();
                // Parallel inner skin has greater curvature on convex bends.
                let skin_radius = radius - p["thickness"] / 2.;
                let k = (skin_radius * skin_radius + 2. * slope * slope - skin_radius * second)
                    .abs()
                    / (skin_radius * skin_radius + slope * slope).powf(1.5);
                max_curvature = max_curvature.max(k);
                arc.push(arc[i] + radius.hypot(slope) * TAU / NA as f64);
            }
            let length = arc[NA];
            for v in &mut arc {
                *v /= length;
            }
            arcs.push(arc);
            lengths.push(length);
        }
        let circumference =
            (lengths.iter().sum::<f64>() - (lengths[0] + lengths[NZ]) / 2.) / NZ as f64;
        let min_circumference = lengths.iter().copied().fold(f64::INFINITY, f64::min);
        let max_circumference = lengths.iter().copied().fold(0f64, f64::max);
        let horizontal_step =
            (8. * 0.02 / max_curvature.max(1e-12)).sqrt() * circumference / max_circumference;
        let mut s = Self {
            arcs,
            vertical: vec![0.],
            circumference,
            min_circumference,
            max_circumference,
            height: 0.,
            horizontal_step,
            axial_height: p["height"],
        };
        for j in 1..=NZ {
            let z0 = p["height"] * (j - 1) as f64 / NZ as f64;
            let z1 = p["height"] * j as f64 / NZ as f64;
            let distance = (0..64)
                .map(|i| {
                    let a = s.point(p, i as f64 / 64., z0, 0.);
                    let b = s.point(p, i as f64 / 64., z1, 0.);
                    ((a[0] - b[0]).powi(2) + (a[1] - b[1]).powi(2) + (a[2] - b[2]).powi(2)).sqrt()
                })
                .sum::<f64>()
                / 64.;
            s.vertical.push(s.vertical[j - 1] + distance);
        }
        s.height = s.vertical[NZ];
        s
    }
    pub fn angle(&self, u: f64, z: f64) -> f64 {
        let t = (z / self.axial_height * NZ as f64).clamp(0., NZ as f64);
        let j = (t.floor() as usize).min(NZ - 1);
        let f = t - j as f64;
        let u = u.rem_euclid(1.);
        let at = |i: usize| self.arcs[j][i] * (1. - f) + self.arcs[j + 1][i] * f;
        let (mut lo, mut hi) = (0, NA);
        while hi - lo > 1 {
            let m = (lo + hi) / 2;
            if at(m) < u {
                lo = m
            } else {
                hi = m
            }
        }
        TAU * (lo as f64 + (u - at(lo)) / (at(hi) - at(lo))) / NA as f64
    }
    pub fn z(&self, v: f64) -> f64 {
        let target = v.clamp(0., 1.) * self.height;
        let hi = self.vertical.partition_point(|x| *x < target).clamp(1, NZ);
        let lo = hi - 1;
        self.axial_height
            * (lo as f64 + (target - self.vertical[lo]) / (self.vertical[hi] - self.vertical[lo]))
            / NZ as f64
    }
    pub fn point(&self, p: &Params, u: f64, z: f64, depth: f64) -> [f64; 3] {
        let a = self.angle(u, z) + p["twist"] * PI / 180. * z / p["height"];
        let r = radius(p, a, z) + depth;
        [r * a.cos(), r * a.sin(), z]
    }
    pub fn counts(&self, p: &Params) -> (f64, usize, usize) {
        let requested = if p.get("cell_pitch").copied().unwrap_or(0.) > 0. {
            p["cell_pitch"]
        } else {
            32. - 26. * p["density"] / 100.
        };
        let pitch = requested.max(2. * p["thickness"] + 2.);
        (
            pitch,
            8.max((self.circumference / pitch).round() as usize),
            3.max((self.height / (pitch * p["cell_aspect"])).round() as usize),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn ripple_adds_cells_and_equalises_surface_distance() {
        let mut p = defaults();
        for k in ["bottom_diameter", "middle_diameter", "top_diameter"] {
            p.insert(k.into(), 160.);
        }
        p.insert("height".into(), 245.);
        p.insert("lobes".into(), 4.);
        p.insert("ripple_depth".into(), 35.);
        p.insert("twist".into(), 45.);
        let s = Surface::new(&p);
        let (_, cols, _) = s.counts(&p);
        assert_eq!(cols, 70);
        assert!((s.circumference - 785.397).abs() < 0.01);
        for col in 0..cols {
            let mut length = 0.;
            for j in 0..64 {
                let a = s.point(&p, (col as f64 + j as f64 / 64.) / cols as f64, 120., 0.);
                let b = s.point(
                    &p,
                    (col as f64 + (j + 1) as f64 / 64.) / cols as f64,
                    120.,
                    0.,
                );
                length += ((a[0] - b[0]).powi(2) + (a[1] - b[1]).powi(2)).sqrt();
            }
            assert!((length - s.circumference / cols as f64).abs() < 0.01);
        }
        p.insert("ripple_depth".into(), 0.);
        assert_eq!(Surface::new(&p).counts(&p).1, 45);
    }
}
